### PR TITLE
Add auth-token to apiClient to allow updating it seamlessly

### PR DESCRIPTION
**Pull request by rohitkum2506**
## Summary

For OH clients to continue with OpenhouseCatalog session after renewed token has been fetched without disrupting the existing ApiClient Connections, we need to provide a way where the auth-token can be added/updated to ApiClient without recreating the ApiClients. 

Since, ApiClient interface allows to update the default header, it can be set on same instance without disrupting any running sessions. ApiClient adds the default headers to request headers while preparing a https(s) request.

Change log:
1. Remove setting header in webClient
2. Add interface in catalog to support updating the auth-token as Authorization header in ApiClient default header

## Changes

- [ ] Client-facing API Changes
- [x] Internal API Changes
- [ ] Bug Fixes
- [x] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [ ] Refactoring
- [ ] Documentation
- [ ] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done

Tested in local docker env and was able to perform basic crud operation using OH catalog with the jar containing changes.
This certifies that the auth-token in ApiClient default header is taking effect. 
Need to test further: When auth-token is updated in ApiClient, it should work seamlessly. Test needs to be done Azkaban.

Screenshot 2024-03-31 at 7 58 31 PM

- [x] Manually Tested on local docker setup. Please include commands ran, and their output.
- [ ] Added new tests for the changes made.
- [ ] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java
Comment (abhisheknath2011) [inline client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java:53]: We need to make this change in all the places where ApiClient is instantiated. For example `JobsApiClientFactory` and `HouseTablesApiClientFactory`. As HTS is internal so token might not be needed. This is one of the reason token was added as part of WebClient. Were you able to see if the the other option of overriding token on WebClient header using ApiClient also works?

Comment (abhisheknath2011) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java:53]: But yeah this approach should be fine as we need to add token through ApiClient and WebClient only allows through mutation which means new instance creation.

Comment (abhisheknath2011) [inline client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java:53]: Is there an existing constant for the Header that can be used in the reactor netty lib?

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (abhisheknath2011) [inline integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:121]:
```suggestion
    if (!token.isEmpty()){
```

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java
Comment (sumedhsakdeo) [inline client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java:53]: Was the secure client not used before in local docker, and hence it worked without the token being set?

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java
Comment (sumedhsakdeo) [inline client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:165]: What code path exercises this?

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (sumedhsakdeo) [reply integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:121]: How does testing done section produce expected result with this bug?

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (sumedhsakdeo) [inline integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:120]: How about private with VisibilityForTesting? Or maybe move this method to unit-test?

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (sumedhsakdeo) [inline integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:100]: Can we revert this refactor and do it in a future PR?

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java
Comment (rohitkum2506) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:165]: WebClient is used in ApiClientFactory Only to create ApiClients. 
Reason to remove is ApiClient also adds the header to webClient headers while preparing the request, so I wanted to avoid case when webClient header is not overridden by ApiClient's default in anyway.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java
Comment (rohitkum2506) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java:53]: Yeah, ApiClient also add these token to webClient headers only while preparing the request, so no need to update in webClient. 
Just that every consumer of webClient will have to do it explicitly.
Updated to add `JobsApiClientFactory` and `HouseTablesApiClientFactory` as well.

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (rohitkum2506) [reply integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:121]: This setToken is intended to be used internal OpenHouse repo to setAuth token fetched from UGI.

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (rohitkum2506) [reply integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:120]: setToken needs visibility to subclass which has to update the auth-token in apiClients headers. Hence its protected.

File: integrations/java/openhouse-java-itest/src/test/java/com/linkedin/openhouse/javaclient/SmokeTest.java
Comment (HotSushi) [inline integrations/java/openhouse-java-itest/src/test/java/com/linkedin/openhouse/javaclient/SmokeTest.java:131]: can we also validate apiClient.getDefaultHeader has the new token after setToken? I'm assuming youll need `VisibilityForTesting`

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (HotSushi) [inline integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:127]: can we pass properties to this method from the initialize method? This method seems to be called only from the initialize method. Helps with code readability and clear relationship with the caller.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java
Comment (abhisheknath2011) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java:53]: Looks like PR is not updated with the latest changes.

File: integrations/java/openhouse-java-itest/src/test/java/com/linkedin/openhouse/javaclient/SmokeTest.java
Comment (rohitkum2506) [reply integrations/java/openhouse-java-itest/src/test/java/com/linkedin/openhouse/javaclient/SmokeTest.java:131]: ApiClient does not expose the defaultHeaders which can be read to validate that tokens are set. 
Although I could verify with debugger that the token is updated in ApiClient, I could not do a get on it to assert.

Overall this test validates that setToken does perform updates on ApiClient but lacks assertions.

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (rohitkum2506) [reply integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:127]: Will wait for Sumedh's comment about keeping this refactoring out.

**Review by sumedhsakdeo — COMMENTED** [review]
I might have missed this PR is still in draft. @rohitkum2506 can you share if it is ready for review?

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (sumedhsakdeo) [reply integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:100]: Yes, please revert. As you point out, the ask is directed towards keeping changes limited to requirements of tokenRefetch. Not sure if this refactor is core to it.

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (sumedhsakdeo) [reply integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java:127]: Let's not mix fixing the code readability and adding token refresh functionality in the same PR. Fine with first doing the code readability PR before the functionality PR, your call.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java
Comment (sumedhsakdeo) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java:53]: @abhisheknath2011 what is the significance of Secure client, certs-only? What do we use in local docker?

Comment (abhisheknath2011) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java]: @sumedhsakdeo Secure client provides way to create one way TLS client and manages HTTP connection pool on the client side. Another significance of Secure client is to provide common way to create REST clients to interact with openhouse. Openhouse catalog and and other parts of the codebase within openhouse/li-openhouse uses Secure client for interacting with openhouse services. In case of local docker secure client is used internally when we interact with openhouse catalog.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java
Comment (abhisheknath2011) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:165]: @rohitkum2506 You can actually do some negative test like not passing the token and observe the errors. You can also also perform some table crud operations etc.

Comment (abhisheknath2011) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:165]: Another way to test is to verify working of jobs/job scheduler.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java
Comment (abhisheknath2011) [inline client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:164]: We need to remove from line 113 as well as test. I think local docker might be using http url.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java
Comment (rohitkum2506) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:165]: Jobs scheduler also uses the same token file to initiate the ApiClient. An independent jobs scheduler tests makes sense since we are changing the ApiClient for Jobs client as well but I think it provides same validations as local Docker tests.

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (rohitkum2506) [reply integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java]: Done, reverted to refactoring. Will being them in after the token refetch is merged.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java
Comment (rohitkum2506) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:164]: ApiClient overrides the Authorization header with its default header while preparing for request. SO, I don't think it should matter it's `http` or `https` url. 
However, for consistency, removed setting headers in `http` request case as well.

**Comment by sumedhsakdeo** [comment]
"Hence, auth-token is now added to ApiClient default headers which is read by ApiClient and added to request header for every client call."

Can you rephrase this sentence in the description? Explain in logical terms what the code change does, not necessarily implementation details.

File: integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java
Comment (rohitkum2506) [reply integrations/java/openhouse-java-runtime/src/main/java/com/linkedin/openhouse/javaclient/OpenHouseCatalog.java]: Removed the refactoring, and limited changes to tokenRefetch only.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java
Comment (rohitkum2506) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/TablesApiClientFactory.java]: All updated now

**Review by abhisheknath2011 — APPROVED** [review]
Thanks Rohit for the changes.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/HousetablesApiClientFactory.java
Comment (sumedhsakdeo) [inline client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/HousetablesApiClientFactory.java:56]:
```suggestion
```
We don't have authentication support in housetables service. Should we remove this?

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/JobsApiClientFactory.java
Comment (sumedhsakdeo) [inline client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/JobsApiClientFactory.java:56]:
```suggestion
```
ditto?

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java
Comment (sumedhsakdeo) [inline client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:110]:
```suggestion
  private WebClient createWebClient(String baseUrl) {
```
We don't need token anymore.

Comment (sumedhsakdeo) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/WebClientFactory.java:110]: Ditto, for createSecureWebClient()

**Review by sumedhsakdeo — COMMENTED** [review]
PR is very clean now to review. Thanks for that @rohitkum2506 :)  ! Some minor comments.

In your docker tests, can you start with `openhouse.token` and update to `tableowner.token` so we see how that is taking effect.

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/JobsApiClientFactory.java
Comment (rohitkum2506) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/JobsApiClientFactory.java:56]: I think it's still worth keeping scope for auth-token support in JobsApiClient, since it's still accessed from external OH managed K8 applications. It's not internal only.

Comment (sumedhsakdeo) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/JobsApiClientFactory.java:56]: Can you point me to the code in jobs service and housetables service where authentication token is intercepted and validated?

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/HousetablesApiClientFactory.java
Comment (rohitkum2506) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/HousetablesApiClientFactory.java]: Done

File: client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/JobsApiClientFactory.java
Comment (rohitkum2506) [reply client/secureclient/src/main/java/com/linkedin/openhouse/client/ssl/JobsApiClientFactory.java]: It's not, removed setting token from JobsClient as well

**Review by sumedhsakdeo — APPROVED** [review]

**Pull request merged**
